torch/futures/__init__.py

```diff
@@ -21,35 +21,101 @@ class Future(torch._C.Future, Generic[T], metaclass=_PyFutureMeta):
+          instructions in the current streams to ensure that further operations
+          enqueued on those streams will be properly scheduled after the async
+          kernels. Once that is done, ``wait()`` will return, even if
+          those kernels are still running. This mode allows for better
+          pipelining and performance, and should be safe as long as one doesn't
+          change streams between the call to ``wait()`` and the time at
+          which the result is used (but it requires explicit manual
+          synchronization when switching streams).
+
+        Args:
+            non_blocking (bool): whether the sync with the GPU kernels should be
+                                 non-blocking.
+
         Returns:
             The value held by this ``Future``. If the function (callback or RPC)
             creating the value has thrown an error, this ``wait`` method will
             also throw an error.
         """
         return super().wait(non_blocking)
 
+    def value(self) -> T:
+        r"""
+        Obtain the value of an already-completed future.
+
+        This method should only be called after a call to :meth:`wait` has
+        completed, or inside a callback function passed to :meth:`then`. In
+        other cases this ``Future`` may not yet hold a value and calling
+        ``value()`` could fail.
+
+        If the value contains tensors that reside on GPUs, then this method will
+        *not* perform any additional synchronization. This should be done
+        beforehand, separately, through a call to :meth:`wait` (except within
+        callbacks, for which it's already being taken care of by :meth:`then`).
+
+        Returns:
+            The value held by this ``Future``. If the function (callback or RPC)
+            creating the value has thrown an error, this ``value()`` method will
+            also throw an error.
+        """
+        return super().value()
+
     # Have to use string annotations because  PEP-0563 is not available in 3.6
     def then(self, callback):  # type: (Callable[[Future[T]], S]) -> Future[S]
         r"""
         Append the given callback function to this ``Future``, which will be run
         when the ``Future`` is completed.  Multiple callbacks can be added to
-        the same ``Future``, and will be invoked in the same order as they were
-        added. The callback must take one argument, which is the reference to
-        this ``Future``. The callback function can use the ``Future.wait()`` API
-        to get the value.
+        the same ``Future``, but the order in which they will be executed cannot
+        be guaranteed (to enforce a certain order consider chaining:
+        ``fut.then(cb1).then(cb2)``). The callback must take one argument, which
+        is the reference to this ``Future``. The callback function can use the
+        :meth:`value` method to get the value.
+
+        If the ``Future``'s value contains tensors that reside on GPUs, the
```

Comment:
As I mentioned above, I don't think this depends on whether the Future contains tensors.